kotlin/src/test/kotlin/com/linecorp/armeria/server/FlowCollectingPublisherTest.kt

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License
+ */
+
+package com.linecorp.armeria.server
+
+import com.linecorp.armeria.internal.common.kotlin.FlowCollectingPublisher
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOf
+import org.assertj.core.api.Assertions.assertThat
+import org.awaitility.Awaitility.await
+import org.junit.jupiter.api.Test
+import org.reactivestreams.Subscriber
+import org.reactivestreams.Subscription
+import java.util.concurrent.BlockingQueue
+import java.util.concurrent.Executors
+import java.util.concurrent.LinkedTransferQueue
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+class FlowCollectingPublisherTest {
+    @Test
+    fun test_shouldCompleteAfterConsumingAllElements() {
+        val queue: BlockingQueue<Any> = LinkedTransferQueue()
+
+        FlowCollectingPublisher(flowOf(1, 2))
+            .subscribe(object : Subscriber<Int> {
+                override fun onSubscribe(s: Subscription) {
+                    queue.add("onSubscribe")
+                    s.request(Long.MAX_VALUE)
+                }
+
+                override fun onNext(t: Int) {
+                    queue.add(t)
+                }
+
+                override fun onError(t: Throwable) {
+                    queue.add("onError") // never reaches here
+                }
+
+                override fun onComplete() {
+                    queue.add("onComplete")
+                }
+            })
+        assertThat(queue.poll(3, TimeUnit.SECONDS)).isEqualTo("onSubscribe")
+        assertThat(queue.poll(3, TimeUnit.SECONDS)).isEqualTo(1)
+        assertThat(queue.poll(3, TimeUnit.SECONDS)).isEqualTo(2)
+        assertThat(queue.poll(3, TimeUnit.SECONDS)).isEqualTo("onComplete")
```

Comment:
We could use `reactor.test.StepVerifier` to verify the outcome of a publisher.

Reply:
I've replaced many of them with `StepVerifier` but I couldn't do so for back-pressure test cases. 😭

Reply:
Backpressure could be tested by implementing Reactive Stream TCK.
https://github.com/line/armeria/blob/6449551d515e2a612b0291f9b72c1c0d368364f2/grpc/src/test/java/com/linecorp/armeria/internal/common/grpc/HttpDeframerTckTest.java#L45

Reply:
Thanks, added a TCK `PublisherVerification` in the test code.
But, the original back-pressure tests should remain there since what I wanted to check was if the back-pressure stops the backing `Flow` from emitting more values.

All `PublisherVerification` tests(but one) passes even though I remove `delegate.whenConsumed().await()`.